File: client/src/page-components/groups/schemas.ts
import { Angle } from "@/types/api";

export interface GroupStats {
  standards_count: number;
  images_count: number;
  annotated_count: number;
  polygons_count: number;
  segment_groups_count: number;
  segments_count: number;
  models_count: number;
}

export interface GroupStandardShort {
  id: string;
  group_id: string;
  name: string | null;
  angle: Angle | null;
  is_active: boolean;
  image_count: number;
  annotated_count: number;
  reference_path: string | null;
  created_at: string;
}

export interface GroupModelShort {
  id: string;
  group_id: string;
  name: string;
  architecture: string;
  version: number;
  epochs: number | null;
  imgsz: number;
  batch_size: number | null;
  num_classes: number | null;
  metrics: Record<string, unknown> | null;
  class_names: string[] | null;
  is_active: boolean;
  trained_at: string | null;
  created_at: string;
}

export interface GroupListItem {
  id: string;
  name: string;
  description: string | null;
  created_at: string;
  stats: GroupStats;
}

export interface GroupDetail {
  id: string;
  name: string;
  description: string | null;
  created_at: string;
  stats: GroupStats;
  standards: GroupStandardShort[];
  active_model: GroupModelShort | null;
}

export interface GroupMutationResponse {
  id: string;
  name: string;
  description: string | null;
  created_at: string;
}

export interface CreateGroupInput {
  name: string;
  description?: string;
}

export interface UpdateGroupInput {
  id: string;
  data: {
    name?: string;
    description?: string;
  };
}

File: server/src/modules/groups/schemas.py
from datetime import datetime
from typing import Annotated, Self
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, StringConstraints, model_validator

Name = Annotated[
    str,
    StringConstraints(
        strip_whitespace=True,
        min_length=1,
        max_length=255,
    ),
]


class GroupStatsResponse(BaseModel):
    standards_count: int = 0
    images_count: int = 0
    annotated_count: int = 0
    polygons_count: int = 0
    segment_groups_count: int = 0
    segments_count: int = 0
    models_count: int = 0


class GroupStandardResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    group_id: UUID
    name: str
    angle: str | None
    is_active: bool
    created_at: datetime
    reference_path: str | None = None
    images_count: int = 0
    annotated_images_count: int = 0


class GroupActiveModelResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    group_id: UUID
    architecture: str
    version: int
    epochs: int | None
    imgsz: int
    batch_size: int | None
    num_classes: int | None
    metrics: dict | None
    class_names: list[str] | None = None
    is_active: bool
    trained_at: datetime | None
    created_at: datetime


class GroupListItemResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    description: str | None
    created_at: datetime
    stats: GroupStatsResponse


class GroupDetailResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    description: str | None
    created_at: datetime
    stats: GroupStatsResponse
    standards: list[GroupStandardResponse] = Field(default_factory=list)
    active_model: GroupActiveModelResponse | None = None


class GroupMutationResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    description: str | None = None
    created_at: datetime


class GroupCreate(BaseModel):
    name: Name
    description: str | None = None


class GroupUpdate(BaseModel):
    name: Name | None = None
    description: str | None = None

    @model_validator(mode="after")
    def validate_not_empty(self) -> Self:
        if not self.model_dump(exclude_unset=True):
            raise ValueError("Необходимо передать хотя бы одно поле")
        return self
